cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,291 @@
+endpoint_index_of_linestring(SizeType const& linestring_idx,
+                             OffsetIterator const& linestring_offsets_begin,
+                             SizeType const& num_linestrings,
+                             SizeType const& num_points)
+{
+  return (linestring_idx == (num_linestrings - 1)
+            ? (num_points)
+            : *(linestring_offsets_begin + linestring_idx + 1)) -
+         1;
+}
+
+/**
+ * @brief Computes shortest distance between @p c and segment ab
+ */
+template <typename T>
+T __device__ point_to_segment_distance_squared(vec_2d<T> const& c,
+                                               vec_2d<T> const& a,
+                                               vec_2d<T> const& b)
+{
+  auto ab        = b - a;
+  auto ac        = c - a;
+  auto l_squared = dot(ab, ab);
+  if (l_squared == 0) { return dot(ac, ac); }
+  auto r  = dot(ac, ab);
+  auto bc = c - b;
+  // If the projection of `c` is outside of segment `ab`, compute point-point distance.
+  if (r <= 0 or r >= l_squared) { return std::min(dot(ac, ac), dot(bc, bc)); }
+  auto p  = a + (r / l_squared) * ab;
+  auto pc = c - p;
+  return dot(pc, pc);
+}
+
+/**
+ * @brief Computes shortest distance between two segments (ab and cd) that
+ * doesn't intersect.
+ */
+template <typename T>
+T __device__ segment_distance_no_intersect_or_colinear(vec_2d<T> const& a,
+                                                       vec_2d<T> const& b,
+                                                       vec_2d<T> const& c,
+                                                       vec_2d<T> const& d)
+{
+  auto dist_sqr = std::min(std::min(point_to_segment_distance_squared(a, c, d),
+                                    point_to_segment_distance_squared(b, c, d)),
+                           std::min(point_to_segment_distance_squared(c, a, b),
+                                    point_to_segment_distance_squared(d, a, b)));
+  return dist_sqr;
+}
+
+/**
+ * @brief Computes shortest distance between two segments.
+ *
+ * If two segments intersect, the distance is 0. Otherwise compute the shortest point
+ * to segment distance.
+ */
+template <typename T>
+T __device__ squared_segment_distance(vec_2d<T> const& a,
+                                      vec_2d<T> const& b,
+                                      vec_2d<T> const& c,
+                                      vec_2d<T> const& d)
+{
+  auto ab    = b - a;
+  auto cd    = d - c;
+  auto denom = det(ab, cd);
+
+  if (denom == 0) {
+    // Segments parallel or collinear
+    return segment_distance_no_intersect_or_colinear(a, b, c, d);
+  }
+
+  auto ac               = c - a;
+  auto r_numer          = det(ac, cd);
+  auto denom_reciprocal = 1 / denom;
+  auto r                = r_numer * denom_reciprocal;
+  auto s                = det(ac, ab) * denom_reciprocal;
+  if (r >= 0 and r <= 1 and s >= 0 and s <= 1) { return 0.0; }
+  return segment_distance_no_intersect_or_colinear(a, b, c, d);
+}
+
+/**
```

Comment:
This shouldn't be in our public docs.

```suggestion
/*
```

Reply:
On that note, I realized cuspatial doesn't have public documentation for cpp.

Reply:
If we do this, my vscode editor will lose highlight to all the entire section. This makes reading a bit harder. Can we use the annotation `@internal` in the section, and set doxygen configuration [`INTERNAL_DOC`](https://www.doxygen.nl/manual/config.html#cfg_internal_docs) to `NO` to hide the section? As described here: https://www.doxygen.nl/manual/commands.html#cmdinternal

Reply:
Sounds like a good solution.

Reply:
Also, in doxygen there's a configuration called `EXCLUDE_PATTERNS`, currently in cudf it includes `*/detail/*` that excludes all APIs if the absolute path contains `detail`.